### PR TITLE
feat(generation): Implement QA Response generator

### LINKED ISSUE #19: feat(generation): Retrieval augmented Question-answering agent
Implement a simple QA agent that uses a retriever to retrieve appropriate context depending on the query and answers the question accordingly.

### PR DESCRIPTION
**Pull request by soumik12345**
Resolves #19 #17 #23

### CODE REVIEW

File: medrag_multi_modal/document_loader/image_loader/base_img_loader.py [hunk 1]
```diff
@@ -61,10 +63,11 @@ async def load_data(
         If a wandb_artifact_name is provided, the processed pages are published to a WandB artifact.
 
         Args:
-            start_page (Optional[int]): The starting page index (0-based) to process. Defaults to the first page.
-            end_page (Optional[int]): The ending page index (0-based) to process. Defaults to the last page.
+            start_page (Optional[int]): The starting page index (0-based) to process.
+            end_page (Optional[int]): The ending page index (0-based) to process.
```
Comment (mratanusarkar) [inline medrag_multi_modal/document_loader/image_loader/base_img_loader.py:67]: why remove the `Defaults to the first/last page.`? that made it intuitive!

Comment (mratanusarkar) [inline medrag_multi_modal/document_loader/image_loader/base_img_loader.py:67]: why remove the `Defaults to the first/last page.`? that made it intuitive!

File: medrag_multi_modal/document_loader/image_loader/base_img_loader.py [hunk 2]
```diff
@@ -99,8 +102,21 @@ async def process_page(page_idx):
...
+        for file in os.listdir(image_save_dir):
+            if file.endswith(tuple(exclude_file_extensions)):
+                os.remove(os.path.join(image_save_dir, file))
+
         if wandb_artifact_name:
-            artifact = wandb.Artifact(name=wandb_artifact_name, type="dataset")
+            artifact = wandb.Artifact(
+                name=wandb_artifact_name,
+                type="dataset",
+                metadata={"loader_name": self.__class__.__name__},
```
Comment (mratanusarkar) [inline medrag_multi_modal/document_loader/image_loader/base_img_loader.py:118]: really helpful!

File: medrag_multi_modal/assistant/medqa_assistant.py [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+            ]
+
+        system_prompt = """
+        You are an expert in medical science. You are given a query and a list of chunks from a medical document.
+        """
+        response = self.llm_client.predict(
+            system_prompt=system_prompt,
+            user_prompt=[query, *retrieved_chunk_texts, *figure_descriptions],
+        )
+        page_numbers = ", ".join([str(int(page_idx) + 1) for page_idx in page_indices])
```
Comment (mratanusarkar) [inline medrag_multi_modal/assistant/medqa_assistant.py:106]: noice!

File: medrag_multi_modal/document_loader/text_loader/base_text_loader.py [hunk 4]
```diff
@@ -131,6 +131,7 @@ async def load_data(
         async def process_page(page_idx):
             nonlocal processed_pages_counter
             page_data = await self.extract_page_data(page_idx, **kwargs)
+            page_data["loader_name"] = self.__class__.__name__
```
Comment (mratanusarkar) [inline medrag_multi_modal/document_loader/text_loader/base_text_loader.py:134]: again! really helpful!

**Review by mratanusarkar — COMMENTED** [review]
left some pointers, rest code LGTM.

also from this [run](https://wandb.ai/ml-colabs/medrag-multi-modal/weave/calls/0192bf31-4335-7fb1-9b32-8b6d95c26cd5), some comments:
- when asked `What is ribosome?`, It tries to structure the retrieved data into points first. while that is useful in engineering and other fields, we need to check if that's the medical approach too.
- there is a small one-line summary towards the end, which should be the ideal one-liner answer. we need to test if this line of "summary" is persistent and always generated, which is the ideal answer to the asked question.
- I like the `Source: Page 35 from Gray's Anatomy` addition in the response footer.
- we need to test for the pages with images, figures, and diagrams, and if the asked question requires diagram reference, then if the current prompt template can handle that?

File: medrag_multi_modal/document_loader/image_loader/base_img_loader.py (hunk 1 above)
Comment (soumik12345) [reply medrag_multi_modal/document_loader/image_loader/base_img_loader.py:67]: This is because `mkdocs` automatically captures defaults.
image

**Comment by soumik12345** [comment]
> * when asked `What is ribosome?`, It tries to structure the retrieved data into points first. while that is useful in engineering and other fields, we need to check if that's the medical approach too.

IMO, this depends on additional context provided with the question. If you simply ask "What is ribosome?", the response is great (with respect to the retrieved chunks). However, if you ask something along the lines of "How can ribosomes be useful in XYZ task?", then surely the LLM would answer is accordingly (again, with respect to the retrieved chunks).

That said, we need to further tune the system prompt and the overall generation workflow for the responses to be useful. We would need feedback from a medical professional along this line.

**Comment by soumik12345** [comment]
> there is a small one-line summary towards the end, which should be the ideal one-liner answer. we need to test if this line of "summary" is persistent and always generated, which is the ideal answer to the asked question.

Yeah, we need to test this. Ideally, we should tune the system prompt to give us the exact kind of answer we want (broken down into points or a big summary). We could also ask it to give both kind of answers and evaluate using an LLM judge.

**Comment by soumik12345** [comment]
> we need to test for the pages with images, figures, and diagrams, and if the asked question requires diagram reference, then if the current prompt template can handle that?

We're already doing that, unroll the latest traces and you can see the `FigureAnnotatorFromPageImage` in action.

**Pull request merged**